Groupby hashmap sampling fully materialized

Benkex opened the pull request:
- Uses Chao1 estimator to get more accurate group size estimations
- Tests for different entry sizes and edge cases

Commit:
GroupBy optimization (switching to sorting if too many groups) 1.0

Commit:
GroupBy: Fix changes

Benkex committed:
GroupBy: Fix "merge results after switch to sort"

And implemented a lazy sampling

Benkex committed:
GroupBy: Fix "no aggregates" case

Benkex committed:
GroupBy: Final changes before PR

Benkex committed:
GroupBy: Revert changes in HashMapBenchmark

Benkex committed:
GroupBy: Revert changes 2.0

Benkex committed:
GroupBy: Fixes

1. Formatting
2. Test passes: GroupByOptimizations
.computeGroupByForSingleIndexScan
3. LOG(DEBUG) --> AD_LOG_DEBUG
ToDo: 21 more test cases, code logic change, ...

Benkex committed:
GroupBy: Fix test cases and code logic

Benkex committed:
GroupBy: Removed troublesome code

Now all tests run

Benkex committed:
GroupBy: Fix formatting

Benkex committed:
GroupBy: Modifications

Benkex committed:
GroupBy SampleFM: Start commit

Sample = Throw some entries into a HashMap.
If the no. of groups exceeds a threshold, we don't use HashMap.
FM = fully materialized

Benkex committed:
GroupBy Sample: Add test and benchmark, refactor

Benkex committed:
GroupBy Sample: Use Chao1 estimator, extend tests, format

Benkex committed:
GroupBy Sample: Refactor, remove benchmark

Replace debug print with a runtimeInfo Detail

Benkex committed:
GroupBy Sample: Remove code for other PR

Benkex committed:
GroupBy Sample: Fix, rem. unnecessary lib

joka921 commented on benchmark/infrastructure/BenchmarkMain.cpp:
```diff
@@ -158,6 +187,13 @@ int main(int argc, char** argv) {
   po::store(po::parse_command_line(argc, argv, options), vm);
   po::notify(vm);
 
+  // Apply GROUP BY sampling params to runtime parameters
+  RuntimeParameters().set<"group-by-sample-enabled">(sampleEnabled);
+  RuntimeParameters().set<"group-by-sample-max-rows">(maxSampleRows);
+  RuntimeParameters().set<"group-by-sample-distinct-ratio">(
+      sampleDistinctRatio);
+  RuntimeParameters().set<"group-by-sample-group-threshold">(groupThreshold);
+
```
As this file is a generic main function for all benchmarks, 
It shouldn't be modified, but all specific changes should be in the specific benchmark.
Also: Please move all the benchmarking stuff into a separate branch/PR s.t. we can focus on the sampling functionality in this branch.

joka921 commented on src/ServerMain.cpp:
```diff
@@ -147,6 +147,28 @@ int main(int argc, char** argv) {
       optionFactory.getProgramOption<"enable-prefilter-on-index-scans">(),
       "If set to false, the prefilter procedures for FILTER expressions are "
       "disabled.");
+  add("group-by-hash-map-enabled",
+      optionFactory.getProgramOption<"group-by-hash-map-enabled">(),
+      "If set, then the GROUP BY operation will use a hash map to store "
+      "intermediate results. This is faster for small groups, but uses more "
+      "memory. If not set, the GROUP BY operation will use a sort-based "
+      "algorithm.");
+  add("group-by-sample-enabled",
+      optionFactory.getProgramOption<"group-by-sample-enabled">(),
+      "If set, then the GROUP BY operation will use a sampling-based hybrid "
+      "algorithm to decide whether to use a hash map or a sort-based algorithm "
+      "for grouping. If not set, the GROUP BY operation will always use a "
+      "sort-based algorithm.");
```
In the final version this probably is an enum
(never-hash-map/always-hash-map/hash-map-via-sample-threshold.

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -320,6 +328,19 @@ sparqlExpression::EvaluationContext GroupByImpl::createEvaluationContext(
+//    join full scan, object count).
+// 3) Otherwise, compute subresult (lazy if possible) and apply hash-map
+// grouping.
+// 4) (PLANNED) If the hash-map grows too large mid-aggregation, buffer
+// remaining data and
+//    switch to sort-based aggregation via mergeSortedTailIntoPartial.
+// 5) (PLANNED) If all optimizations fail or threshold crossed, fallback to
+// standard doGroupBy (sort+group-by).
```
Planned features shouldn't already be in the comment for this PR.

joka921 commented on src/engine/GroupByImpl.h:
```diff
@@ -591,9 +591,49 @@ class GroupByImpl : public Operation {
-// _____________________________________________________________________________
+  FRIEND_TEST(GroupBySamplingTest, belowThreshold);
+  FRIEND_TEST(GroupBySamplingTest, aboveThreshold);
+  FRIEND_TEST(GroupBySamplingTest, edgeCaseEmptyInput);
+  FRIEND_TEST(GroupBySamplingTest, edgeCaseAllUnique);
+  FRIEND_TEST(GroupBySamplingTest, edgeCaseAllSame);
+
+ private:
+  // Check via sampling if the hash-map path should be skipped due to too many
+  // groups
+  bool shouldSkipHashMapGrouping(const IdTable& table) const;
+  bool shouldSkipHashMapGroupingDynamic(
+      const std::shared_ptr<const Result>& subresult) const;
+  
+  struct RowKey {
+    const IdTable* table;
+    size_t rowIndex;
+    const std::vector<ColumnIndex>* groupByCols;
+
+    // We store RowKeys in absl::flat_hash_set. Therefore, we need to
+    // implement a Hash method and an equality operator.
+
+    template <typename H>
+    friend H AbslHashValue(H h, const RowKey& key) {
+      for (const auto& colIdx : *key.groupByCols) {
+        h = H::combine(std::move(h), (*key.table)(key.rowIndex, colIdx));
+      }
+      return h;
+    }
+
+    bool operator==(const RowKey& other) const {
+      for (const auto& colIdx : *groupByCols) {
+        if ((*table)(rowIndex, colIdx) != (*other.table)(other.rowIndex, colIdx)) {
+          return false;
+        }
+      }
+      return true;
+    }
+  };
+
```
As this functionality only depends on the `Table` and the `groupByCols`, it can be 
externalized as a separate class (the GroupBy is already long enough). 
Call it something like `GroupByAlgorithmSelector` or something, and implement it separately (it only works on IdTables and grouped columns).

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -348,14 +369,26 @@ Result GroupByImpl::computeResult(bool requestLaziness) {
+                              "falling back to sort-based grouping.");
+      useHashMapOptimization = false;
+    } else {
+      subresult = child->getResult(true);
```
You have already populated the `subresult`, no need to do it again.

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -1642,3 +1675,70 @@ std::unique_ptr<Operation> GroupByImpl::cloneImpl() const {
+  if (maxSampleRows != 0) {
+    sampleSize = std::min(sampleSize, maxSampleRows);
+  }
+  // Reservoir-sample indices only
+  std::vector<size_t> indices(sampleSize);
+  for (size_t i = 0; i < sampleSize; ++i) indices[i] = i;
+  std::mt19937_64 gen{42};
+  for (size_t i = sampleSize; i < totalSize; ++i) {
+    std::uniform_int_distribution<size_t> dist(0, i);
+    size_t randIdx = dist(gen);
+    if (randIdx < sampleSize) {
+      indices[randIdx] = i;
+    }
+  }
```
1. We have helpers in `util/Random` that make the randomness much easier to use, please use them.
2.  sampling your indices is really strange, why do you not sample a random index from [0, size of the input] and then just add it to the sampled indices?

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -1642,3 +1675,70 @@ std::unique_ptr<Operation> GroupByImpl::cloneImpl() const {
+    }
+  }
+  // Use absl::flat_hash_map to count frequencies for Chao1 estimator
+  absl::flat_hash_map<RowKey, size_t> groupCounts;
```
use the alias (+ the include) `ad_utility::HashMap`. instead of the absl types directly.

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -1642,3 +1675,70 @@ std::unique_ptr<Operation> GroupByImpl::cloneImpl() const {
+  for (const auto& var : _groupByVariables) {
+    groupByCols.push_back(varCols.at(var).columnIndex_);
+  }
+  for (size_t idx : indices) {
+    RowKey key{&table, idx, &groupByCols};
+    ++groupCounts[key];
+  }
```
You can do this directly for each of the sampled indices, no need to materialized them in a vector first.

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -1642,3 +1675,70 @@ std::unique_ptr<Operation> GroupByImpl::cloneImpl() const {
+  size_t dObs = groupCounts.size();
+  size_t f1 = 0, f2 = 0;
+  for (auto& [key, cnt] : groupCounts) {
+    if (cnt == 1)
+      ++f1;
+    else if (cnt == 2)
+      ++f2;
+  }
```
Am I seeing this correctly, this estimator only uses groups that appear once or twice. Just checking that this is correct. (now that I think about it, it looks reasonabley, because "frequent groups" do not contribute much to the number of groups.

joka921 commented on src/engine/GroupByImpl.h:
```diff
@@ -591,9 +591,49 @@ class GroupByImpl : public Operation {
+  // Check via sampling if the hash-map path should be skipped due to too many
+  // groups
+  bool shouldSkipHashMapGrouping(const IdTable& table) const;
+  bool shouldSkipHashMapGroupingDynamic(
+      const std::shared_ptr<const Result>& subresult) const;
```
Have you already looked (etc. at the Service implementation, which also has to store parts of a lazy result), 
on how you can do this estimation for lazy results? This would be the most important feature here, as the subresults are often lazy, and it is only the sort that makes them non-lazy.

joka921 commented on test/util/IndexTestHelpers.cpp:
```diff
@@ -306,8 +306,7 @@ QueryExecutionContext* getQec(TestIndexConfig c) {
         std::make_unique<QueryExecutionContext>(
-            *index_, cache_.get(), makeAllocator(MemorySize::megabytes(100)),
-            SortPerformanceEstimator{});
+            *index_, cache_.get(), makeAllocator(), SortPerformanceEstimator{});
```
Is that change here important, and if yes, for what?

Benkex replied:
I removed memory limit to be able to use GroupByHashMapBenchmark (I got a limit violation every time). But this is not relevant for the sampling mechanism, so I will move this to a separate PR for benchmarking as well.

joka921 commented on test/engine/GroupBySamplingTest.cpp:
```diff
@@ -0,0 +1,172 @@
+#include <string>
+#include <vector>
+
+#include "../../src/engine/GroupByImpl.h"
+#include "../../src/engine/QueryExecutionContext.h"
+#include "../../src/engine/QueryExecutionTree.h"
+#include "../../src/engine/Result.h"
+#include "../../src/engine/Values.h"
+#include "../../src/engine/idTable/IdTable.h"
+#include "../../src/global/RuntimeParameters.h"
+#include "../../src/parser/GraphPatternOperation.h"
```
Use absolut include paths for everything inside `src` (according to the style that all the other files use).

joka921 reviewed (changes requested):
A first round of reviews, fix everything that you can easily fix yourself, answer my questions, and let me know where there's trouble with the more complex questions.

Benkex committed:
GroupBy Sample: New class, refactor

Benkex committed:
GroupBy Sample: Format

Benkex committed:
GroupBy Sampe: Modifications

Benkex committed:
GroupBy Sample: Remove benchmark

Memory limit and other benchmark related changes
will be in another branch

Benkex committed:
GroupBy Sample: Remove obsolete imports

Benkex committed:
GroupBy Sample: Fixed tests

Benkex committed:
GroupBy Sample: Separate functionality, add timing

Benkex committed:
GroupBy Sample: Separate helpers from test

joka921 reviewed (changes requested):
Some comments we have to take into account.

Benkex committed:
GroupBy Sample: Fix trial 1

IN PROGRESS
We want to solve the problem that the subresult is discarded, and computed again.
